src/main/scala/quasar/sbtdatasource/AssemblePlugin.scala

```diff
@@ -0,0 +1,205 @@
+import java.lang.{Exception, Runtime, SuppressWarnings}
+import java.nio.file.{Path, Files, SimpleFileVisitor, FileVisitResult}
+import java.nio.file.attribute.BasicFileAttributes
+import java.nio.file.StandardCopyOption.REPLACE_EXISTING
+import java.nio.file.StandardOpenOption.CREATE_NEW
+
+import cats.Parallel
+import cats.data.ValidatedNel
+import cats.effect.Sync
+import cats.instances.list._
+import cats.instances.string._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.foldable._
+import cats.syntax.functor._
+import cats.syntax.parallel._
+import cats.syntax.traverse._
+
+import coursier._
+import coursier.util.Schedulable
+
+import io.circe.Json
+
+import sbt.ModuleID
+
+/** Assemble a datasource plugin tarball, returning the path to the artifact. */
+object AssemblePlugin {
+  @SuppressWarnings(Array(
+    "org.wartremover.warts.NonUnitStatements",
+    "org.wartremover.warts.ToString"))
+  def apply[F[_]: Schedulable: Sync, G[_]](
+      dsName: String,
+      dsVersion: String,
+      dsDependencies: Seq[ModuleID],
+      dsJar: Path,
+      quasarVersion: String,
+      dstDir: Path)(
+      implicit P: Parallel[F, G])
+      : F[Path] = {
+
+    // We assemble every component of the final tarball in this folder
+    val buildDir = dstDir.resolve("plugin-build")
+
+    // Will contain the dependencies and the plugin jar itself
+    val pluginDir = buildDir.resolve(dsName)
+
+    // The jar file's own path.
+    val jarPath = pluginDir.resolve(dsJar.getFileName)
+
+    // the datasource jar's path relative to the build dir
+    // included in the generated .plugin file to let quasar
+    // know where to load it from.
+    val relativeJarPath = buildDir.relativize(jarPath)
+
+    // the path to the generated .plugin file.
+    val pluginPath = buildDir.resolve(s"$dsName.plugin")
+
+    // start coursier on resolving all of the datasource's
+    // dependencies, *except* for quasar. quasar and its
+    // dependencies are already present in the user's
+    // `plugins` folder.
+    val resolution = Resolution(dsDependencies.map(moduleIdToDependency).toSet)
+
+    val cache = Cache.fetch[F](buildDir.toFile, CachePolicy.Update)
+
+    type OrFileErrors[A] = ValidatedNel[FileError, A]
+
+    for {
+      _ <- Sync[F] delay {
+        if (Files.exists(buildDir)) {
+          Files.walkFileTree(buildDir, RecursiveDeleteVisitor)
+        }
+      }
+
+      _ <- Sync[F].delay(Files.createDirectories(pluginDir))
+
+      _ <- Sync[F].delay(println("Fetching artifacts with coursier..."))
+
+      copyJar = Sync[F].delay(Files.copy(dsJar, jarPath, REPLACE_EXISTING))
+
+      fetchJarFiles = for {
+        metadata <- ResolutionProcess(resolution).run(
+          // we don't use ~/.ivy2/local here because
+          // I've heard that coursier doesn't copy
+          // from local caches into other local caches.
+          Fetch.from(Seq(MavenRepository("https://repo1.maven.org/maven2")), cache))
```

Comment:
Won't this fail to actually find artifacts? Like, say, quasar? Don't we need to include our public bintray?

Reply:
Yeah, we should, I'll include it. However, we explicitly avoid pulling in quasar (which is why we have the `datasourceDependencies` key), I was a bit fuzzy on that part. Will the plugin use the already loaded quasar classes when it is loaded in at runtime, or does it need to include its own?

If the latter, then we don't need the extra setting key and can just use `libraryDependencies`

Reply:
Definitely the former. Classloaders inherit from the parent, which in this case is quasar. But yeah, you're right that we might not necessarily need the public bintray. Certainly doesn't hurt to include it.